tests/shakenz_test.go

```diff
@@ -0,0 +1,279 @@
+				if connection.Station != sensor.Station {
+					continue
+				}
+				if connection.Location != sensor.Location {
+					continue
+				}
+
+				// run through the deployed dataloggers
+				for _, datalogger := range deployedDataloggerList {
+					// must still be deployed
+					if time.Now().After(datalogger.End) {
+						continue
+					}
+
+					// must also be at the connection place and role
+					if datalogger.Place != connection.Place {
+						continue
+					}
+					if datalogger.Role != connection.Role {
+						continue
+					}
+
+					// run through response stages for this equipment pair
+					for _, response := range resp.Streams(datalogger.Model, sensor.Model) {
+						if stream.SamplingRate != response.SampleRate {
+							continue
+						}
+						for _, channel := range response.Channels(stream.Axial) {
+
+							source := strings.Join([]string{network, stream.Station, stream.Location, channel}, "_")
+							if !match.MatchString(source) {
+								continue
+							}
+
+							// add current stream
+							if q, ok := Q[int(response.SampleRate)]; ok {
+								streams.Streams[source] = &delta.ShakeNZStream{
+									Sensor:     sensor.Model,
+									Datalogger: datalogger.Model,
+									StreamId:     source,
+									Network:    network,
+									Station:    stream.Station,
+									Location:   stream.Location,
+									Channel:    channel,
+									SampleRate: response.SampleRate,
+									Gain:       response.Gain(),
+									Q:          q,
+								}
+							}
+						}
+
+					}
+				}
+			}
+		}
+	}
+
+	// get a map of station and site location then we can add longitude and latitude to the config.
+	s, err := stations()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for k, v := range streams.Streams {
+		if _, ok := s.Stations[v.Station]; !ok {
+			t.Errorf("no station for stream %s can't find longitude latitude", k)
+		}
+		if _, ok := s.Stations[v.Station].Sites[v.Location]; !ok {
+			t.Errorf("no site for stream %s can't find longitude latitude", k)
+		}
+
+		v.Longitude = s.Stations[v.Station].Sites[v.Location].Point.Longitude
+		v.Latitude = s.Stations[v.Station].Sites[v.Location].Point.Latitude
+
+		v.Source = v.Network + "." + v.Station + "." + v.Location
+
+		if len(v.Channel) != 3 {
+			t.Errorf("chan not len 3 for %s", k)
+		}
+
+		// TODO should we also use components 1,2,3?
+		switch strings.ToUpper(string(v.Channel[2])) {
+		case "Z":
+			v.Vertical = true
+			streams.Streams[k] = v
+		case "N", "E":
+			v.Horizontal = true
+			streams.Streams[k] = v
+		default:
+			delete(streams.Streams, k)
+		}
+	}
+
+	if len(streams.Streams) == 0 {
+		t.Fatal("got zero length streams for shakenz config.")
+	}
+
+	b, err := proto.Marshal(&streams)
+	if err != nil {
+		t.Fatal(err)
+
+	}
+
+	if err := os.MkdirAll(apiDir + "/config", 0777); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := ioutil.WriteFile(apiDir + "/config/shakenz-slink.pb", b, 0644); err != nil {
+		t.Fatal(err)
+	}
+}
```

Comment:
That's a pretty long and deeply nested test but I can see why.  The comments make it easy to see the intent and I don't see much point in splitting it up into different functions since it's only used here.